packages/core/src/useOnGetState.tsx

```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
+import NavigationBuilderContext from './NavigationBuilderContext';
+import { NavigationState } from './types';
+import NavigationRouteContext from './NavigationRouteContext';
+
+export default function useOnGetState({
+  getStateForRoute,
+  getState,
+}: {
+  getStateForRoute: (routeName: string) => NavigationState;
+  getState: () => NavigationState;
+}) {
+  const { addStateGetter } = React.useContext(NavigationBuilderContext);
+  const route = React.useContext(NavigationRouteContext);
+  const key = route ? route.key : 'root';
+
+  const getter = React.useCallback(
+    () => ({
+      ...getState(),
+      routes: getState().routes.map(route => ({
+        ...route,
+        state: route.state ? getStateForRoute(route.key) : route.state,
```

Comment:
`route.state` might be undefined, but there can still be a child navigator. Change to:

```suggestion
        state: getStateForRoute(route.key),
```

We'll need to change `getStateForRoute` to handle non-existent keys and return `undefined`